Implements a variable deposit base calculation for EPM signed submissions

gpestana opened the pull request:
This PR implements a generic BaseDeposit calculation for signed submissions, based on the size of the submission queue. 

It adds a new associated type to EPM's config, `type SignedDepositBase`, that implements `Convert<usize, BalanceOf<T>>`, which is used to calculate the base deposit for signed submissions based on the size of the signed submissions queue.

`struct GeometricDepositBase<Balance, Fixed, Inc>` implements the convert trait so that the deposit value increases as a geometric progression. The deposit base is calculated by `deposit_base = fixed_deposit_base * (1 + increase_factor)^n`, where `n` is the term of the progression (i.e. the number of signed submissions in the queue). `Fixed` and `Inc` generic params are getters for `Balance` and `IncreaseFactor` to compute the geometric progression. If `IncreaseFactor =  0`, then the signed deposit is constant and equal to `Fixed` regardless of the size of the queue. 

polkadot companion: https://github.com/paritytech/polkadot/pull/7140

Closes https://github.com/paritytech/srlabs_findings/issues/189

gpestana committed:
Implements a variable deposit base calculation in EPM

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2718478

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2718477

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2718492

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2718488

gupnik committed:
Adds force_origin support (#13845)

* Adds force_origin support

* Moves a couple of tests to showcase v2 with force_origin

* Adds remaining tests

* adds documentation

* minor

* adds test for invalid origin

* ".git/.scripts/commands/fmt/fmt.sh"

* updates param to use MaxCalls

* Fixes compilation error

* Updates doc comment

* Fixes test outputs

* Fixes test output

* ".git/.scripts/commands/fmt/fmt.sh"

---------

Co-authored-by: command-bot <>

muharem committed:
Vote locks for all reasons except RESERVE (#13914)

* Vote locks tip

* except reserve

* reason for delegate

* fix tests

---------

Co-authored-by: parity-processbot <>

alvicsam committed:
[ci] Update buildah command and version (#13989)

dependabot[bot] committed:
Bump enumflags2 from 0.7.5 to 0.7.7 (#13995)

Bumps [enumflags2](https://github.com/meithecatte/enumflags2) from 0.7.5 to 0.7.7.
- [Release notes](https://github.com/meithecatte/enumflags2/releases)
- [Commits](https://github.com/meithecatte/enumflags2/compare/v0.7.5...v0.7.7)

---
updated-dependencies:
- dependency-name: enumflags2
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

bkchr committed:
frame-support-procedural: Fix detection of the tuples feature (#13996)

We didn't had the tuples features declared for the `frame-support-procedural` crate and thus, it
could not properly detect that the feature was already enabled.

bkchr committed:
pallet-democracy: Do not request the proposal when scheduling (#13827)

The requesting of the proposal is actually done now in `pallet-scheduler`.

Fixes: https://github.com/paritytech/substrate/issues/13534

alvicsam committed:
[ci] Add message to cargo-deny (#14001)

* [ci] Add message to cargo-deny

* fix cargo-deny-licenses

yjhmelody committed:
refactor(cli): Make some run params reusable (#13870)

* refactor: extract TelemetryParams and PrometheusParams

* improve run_cmd docs

* extract `RuntimeParams`

* use `normalize`

* keep params types same style

* improve `max_runtime_instances`

* fmt

* add license and improve code

* Update client/cli/src/params/runtime_params.rs

Co-authored-by: Bastian Köcher <git@kchr.de>

---------

Co-authored-by: Bastian Köcher <git@kchr.de>

pgherveou committed:
contracts Add LOG_TARGET constant (#14002)

* contracts Add LOG_TARGET constant

* Update frame/contracts/src/lib.rs

Co-authored-by: Sasha Gryaznov <hi@agryaznov.com>

---------

Co-authored-by: Sasha Gryaznov <hi@agryaznov.com>

koushiro committed:
frame-support: migrate some tests from `decl_*` macros to the new `pallet` macros (#12401)

* frame-support: migrate some tests from decl macros to new pallet attribute macros

* Remove useless type alias

* Remove useless type alias

* Work around for rust issue 52234

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Fix tests

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Use polkadot-compatible paste version

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Fix crate access and add tests

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Typo

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

---------

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>
Co-authored-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>
Co-authored-by: parity-processbot <>

gupnik committed:
Removes ReportsByKindIndex (#13936)

* Removes ReportsByKind

* Minor build fixes

* adds migration

* Addresses review comment

* Uses clear but weight check fails

* Uses unique

* Updates test to commit the change before migration

* Uses reads_writes

* ".git/.scripts/commands/fmt/fmt.sh"

* Fixes build

* Addresses review comments

* ".git/.scripts/commands/fmt/fmt.sh"

* fixes typo

---------

Co-authored-by: command-bot <>

gilescope committed:
Allow missing docs for autogen weights. (#14011)

gpestana committed:
Implements geometric increase of base deposit based on the number of elements in the queue; adds tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2733705

agryaznov committed:
[contracts] Port host functions to Weight V2 and storage deposit limit (#13565)

* added [unstable][seal2] call()

* updated test to cover new seal_call proof_limit

* docs updated

* add [seal2][unstable] instantiate() and test

* add [seal2][unstable] weight_to_fee() + docs and test

* add [seal2][unstable] gas_left() + docs and test

* update benchmarks

* add DefaultDepositLimit to pallet Config

* specify deposit limit for nested call

add test for nested call deposit limit

save: separate deposit limit for nested calls

* specify deposit limit for nested instantiate

save: works with test

cleaned up debugging outputs

* update benchmarks

* added missing fixtures

* fix benches

* pass explicit deposit limit to storage bench

* explicit deposit limit for another set_storage bench

* add more deposit limit for storage benches

* moving to simplified benchmarks

* moved to simplified benchmarks

* fix seal_weight_to_fee bench

* fix seal_instantiate benchmark

* doc typo fix

* default dl for benchmarking

more dl for tests

dl for tests to max

deposit_limit fix in instantiate bench

fix instantiate bench

fix instantiate benchmark

fix instantiate bench again

remove dbg

fix seal bench again

fixing it still

seal_instantiate zero deposit

less runs to check if deposit enough

try

try 2

try 3

try 4

* max_runtime_mem to Schedule limits

* add default deposit limit fallback check to test

* weight params renaming

* fmt

* Update frame/contracts/src/benchmarking/mod.rs

Co-authored-by: PG Herveou <pgherveou@gmail.com>

* prettify inputs in tests

* typestate param refactored

---------

Co-authored-by: PG Herveou <pgherveou@gmail.com>

koushiro committed:
sp-core: remove useless bounded module (#13865)

agryaznov committed:
fix a test (#14021)

gavofyork committed:
Various minor fixes (#13945)

* Fix: Incorrect implementation of can_reserve check

* Fix: Incorrect migration of consumer counting for existing accounts with frozen amounts

* Fix: Inconsistent implementation between assets can_deposit and new_account

* Fixes

* Fixes

* Another fix

* Update tests.rs

* Update fungible_tests.rs

* Use `can_accrue_consumers` in the body of `can_inc_consumer`

---------

Co-authored-by: Keith Yeung <kungfukeith11@gmail.com>
Co-authored-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>
Co-authored-by: parity-processbot <>

pgherveou committed:
contracts Add storage_deposit test (#14003)

* contracts Add storage_deposit test

* fix comments

* PR comments

liamaharon committed:
remote-externalities: batch insert key/values (#14004)

* add batch inserting into remote externalities

* use into_iter

* remove redundant comment

* redundant batch insert

* avoid extra vec allocations

liamaharon committed:
collective pallet: sort genesis members and enforce max len constraint (#13988)

* insert members in sorted order

* improve variable name

* enforce genesis members length constraint

bkchr committed:
sc-network-sync: Improve error reporting (#14025)

Before this wasn't printing a warning for `VerificationFailed` when there wasn't a peer assigned to
the error message. However, if there happens an error on importing the state there wouldn't be any
error. This pr improves the situation to also print an error in this case. Besides that there are
some other cleanups.

gpestana committed:
Finishes implementation and test

pgherveou committed:
contracts Fix store-call test path (#14028)

gpestana committed:
More docs

yjhmelody committed:
chore(cli): make cli display docs correctly (#14017)

* chore(cli): make cli display docs correctly

* ".git/.scripts/commands/fmt/fmt.sh"

---------

Co-authored-by: command-bot <>

gpestana committed:
fixes docs

kianenigma committed:
improve staking interface methods (#14023)

* improve staking interface methods

* fix

* fix

* fix build

* restart

* fix

---------

Co-authored-by: parity-processbot <>

liamaharon committed:
try-runtime-cli: improve ci stability  (#14030)

* hardcode 1 thread

* ci stability

* fix comment

* improve comment

* kick ci

* kick ci

* update expect message

* improve comment

* kick ci

* kick ci

* configure threads with env var

* fix threads env var

* fix threads env var

ggwpez committed:
FRAME: inherited call weight syntax (#13932)

* First approach on pallet::call_weight

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Use attr on pallet::call instead

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Ui tests

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Rename to weight(prefix = ...))

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Simplify to #[pallet::call(weight(T))]

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Add stray token error

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Cleanup

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Migrate remaining pallets

Using script from https://github.com/ggwpez/substrate-scripts/blob/e1b5ea5b5b4018867f3e869fce6f448b4ba9d71f/frame-code-migration/src/call_weight.rs

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Try to add some docs

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Revert "Migrate remaining pallets"

Lets do this as a follow-up, I dont want to bloat this small MR.

This reverts commit 331d4b42d72de1dacaed714d69166fa1bc9c92dd.

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Renames

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Review fixes

Co-authored-by: Sam Johnson <sam@durosoft.com>

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Test weights

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Update UI tests

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Update frame/support/procedural/src/pallet/parse/mod.rs

Co-authored-by: Muharem Ismailov <ismailov.m.h@gmail.com>

* Remove old code

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Update docs

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

---------

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>
Co-authored-by: parity-processbot <>
Co-authored-by: Muharem Ismailov <ismailov.m.h@gmail.com>

gpestana committed:
Implements `try_state` hook in elections and EPM pallets (#13979)

* Adds try_state hook to elections pallets

* Addresses PR review comments

Co-authored-by: Liam Aharon <liam.aharon@hotmail.com>

* remove unecessary println

* ensures try-runtime does not mutate storage

* Addresses PR comments

* Fixes snapshot invariant checks; simplifies test infra

---------

Co-authored-by: Liam Aharon <liam.aharon@hotmail.com>
Co-authored-by: parity-processbot <>

dependabot[bot] committed:
Bump wasmtime from 6.0.1 to 6.0.2 (#14037)

Bumps [wasmtime](https://github.com/bytecodealliance/wasmtime) from 6.0.1 to 6.0.2.
- [Release notes](https://github.com/bytecodealliance/wasmtime/releases)
- [Changelog](https://github.com/bytecodealliance/wasmtime/blob/main/docs/WASI-some-possible-changes.md)
- [Commits](https://github.com/bytecodealliance/wasmtime/compare/v6.0.1...v6.0.2)

---
updated-dependencies:
- dependency-name: wasmtime
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

bkchr committed:
Improve contribution guidelines (#13902)

ggwpez committed:
CI: Remove crate publish check (#14044)

* Fix frame-support feature

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

* Remove publishing script

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

---------

Signed-off-by: Oliver Tale-Yazdi <oliver.tale-yazdi@parity.io>

yjhmelody committed:
fix(in_mem): fix the clone logic (#14038)

pgherveou committed:
Contracts: runtime_call and storage_deposit (#13990)

* wip

* add comments

* fix comment

* comments

* comments

* PR comment

* field orders

* Update frame/contracts/src/tests.rs

* Update frame/contracts/fixtures/call_runtime_and_call.wat

Co-authored-by: Sasha Gryaznov <hi@agryaznov.com>

* Apply suggestions from code review

Co-authored-by: Sasha Gryaznov <hi@agryaznov.com>

* Apply suggestions from code review

Co-authored-by: Sasha Gryaznov <hi@agryaznov.com>

* Update frame/contracts/src/tests.rs

Co-authored-by: Sasha Gryaznov <hi@agryaznov.com>

* Validate fees of failed call

* Update frame/contracts/src/tests.rs

* Update frame/contracts/src/tests.rs

* Update frame/contracts/src/tests.rs

* bubble up refund error

* rename fixture file

---------

Co-authored-by: Sasha Gryaznov <hi@agryaznov.com>
Co-authored-by: parity-processbot <>

bkchr committed:
Fix bags-list tests execution (#14047)

The `test` feature isn't propagated to dependencies in Rust, so we should just enable the method as
well for `std`.

Ank4n commented on frame/election-provider-multi-phase/src/signed.rs:
```diff
@@ -349,6 +351,36 @@ impl<T: Config> SignedSubmissions<T> {
+	/// submissions.
+	///
+	/// The nth term is obtained by calculating `base * (1 + increase_factor)^nth`. Example: factor
+	/// 5, with initial deposit of 1000 and 10% of increase factor is 1000 * (1 + 1.1)^5.
```
Is the calculation formula correct or may be there is a typo?
```suggestion
	/// 5, with initial deposit of 1000 and 10% of increase factor is 1000 * (1 + 0.1)^5.
```

Ank4n commented on frame/staking/src/lib.rs:
```diff
@@ -159,7 +159,7 @@
 //! use pallet_staking::{self as staking};
 //!
-//! #[frame_support::pallet]
+//! #[frame_support::pallet(dev_mode)]
```
I guess you are trying to get rid of warnings in the doc tests? 😄

gpestana replied:
Yes :D

Ank4n commented on bin/node/runtime/src/lib.rs:
```diff
@@ -746,7 +747,9 @@ impl pallet_election_provider_multi_phase::Config for Runtime {
 	type SignedRewardBase = SignedRewardBase;
-	type SignedDepositBase = SignedDepositBase;
+	type SignedDepositBase = GeometricDepositBase<Runtime>;
+	type SignedFixedDepositBase = SignedDepositBase;
```
Is `SignedFixedDepositBase` used anymore? Can we get rid of it? I believe if we want to use fixed deposit base, we should provide `SignedDepositBase ` with a `FixedDepositBase<>` instead of `GeometricDepositBase<>`?

Ank4n commented on bin/node/runtime/src/lib.rs:
```diff
@@ -621,6 +621,7 @@ parameter_types! {
 	// signed config
 	pub const SignedRewardBase: Balance = 1 * DOLLARS;
 	pub const SignedDepositBase: Balance = 1 * DOLLARS;
```
```suggestion
	pub const SignedFixedDepositBase: Balance = 1 * DOLLARS;
```

Ank4n commented on frame/election-provider-multi-phase/src/signed.rs:
```diff
@@ -349,6 +351,36 @@ impl<T: Config> SignedSubmissions<T> {
+	/// Increase factor of the geometric progression, as a percentage.
+	///
+	/// The nth progression term will be `IncreaseFactor`% larger than the nth-1 term.
+	type IncreaseFactor = T::SignedDepositBaseIncreaseFactor;
```
I think a small improvement would be to add `FixedDepositBase` as an associated type of this trait along with `IncreaseFactor`.

Ank4n commented on frame/election-provider-multi-phase/src/lib.rs:
```diff
@@ -673,6 +675,16 @@ pub mod pallet {
 
+		/// Something that calculates the signed deposit base based on the signed submissions queue
+		/// size.
+		type SignedDepositBase: SignedDepositBase<BalanceOf<Self>>;
```
No strong opinion but would it make sense to name the type and trait something along the lines of `DepositBaseCalculator` or something to differentiate it from two other very similar types `SignedDepositBaseIncreaseFactor` and `SignedFixedDepositBase `?

gpestana replied:
Yes, I do agree that `SignedDepositBase` is too similar to the other `SignedDeposit*` types. I like the  `DepositBaseCalculator`, I'll update the name.

- [ ] Rename `SignedDepositBase` -> `DepositBaseCalculator`

Ank4n commented on frame/election-provider-support/src/traits.rs:
```diff
@@ -123,3 +124,12 @@ where
 }
+
+/// A type that calculates the deposit base for signed election submissions.
+pub trait SignedDepositBase<Balance> {
```
Nit: Does this trait need to be specific to Signed submissions? Or could it may be a bit more generic that calculates fee in a geometrical sequence based on queue size?

gpestana replied:
Refactored this trait to be more generic and to calculate any deposit based on the queue size.

rcny committed:
CI: migrate to Google Cloud (#13994)

lexnv committed:
rpc: Use the blocks pinning API for chainHead methods (#13233)

* rpc/chain_head: Add backend to subscription management

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/chain_head: Pin blocks internally and adjust testing

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* client/in_mem: Reference for the number of pinned blocks

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/tests: Check in-memory references to pinned blocks

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/chain_head: Fix clippy

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/chain_head: Remove unused comment

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/chain_head: Place subscription handle under `Arc` and unpin blocks on drop

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/tests: Check all pinned blocks are unpinned on drop

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* Apply suggestions from code review

Co-authored-by: Bastian Köcher <git@kchr.de>

* Update client/rpc-spec-v2/src/chain_head/subscription.rs

Co-authored-by: Bastian Köcher <git@kchr.de>

* rpc/tests: Retry fetching the pinned references for CI correctness

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* client/service: Use 512 as maximum number of pinned blocks

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chain_head: Fix merging conflicts

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/chain_head: Adjust subscriptions to use pinning API

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/chain_head/tests: Test subscription management

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/chain_head: Adjust chain_head follow to the new API

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/chain_head: Adjust chain_head.rs to the new API

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/chain_head/tests: Adjust test.rs to the new API

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* client/builder: Use new chainHead API

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/chain_head: Fix documentation

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* rpc/chain_head: Fix clippy

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* client/in_mem: ChainHead no longer uses `in_mem::children`

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* Update client/rpc-spec-v2/src/chain_head/subscription.rs

Co-authored-by: Sebastian Kunert <skunert49@gmail.com>

* Update client/rpc-spec-v2/src/chain_head/subscription.rs

Co-authored-by: Sebastian Kunert <skunert49@gmail.com>

* Update client/rpc-spec-v2/src/chain_head/subscription.rs

Co-authored-by: Sebastian Kunert <skunert49@gmail.com>

* Update client/rpc-spec-v2/src/chain_head/subscription.rs

Co-authored-by: Sebastian Kunert <skunert49@gmail.com>

* chain_head: Add block state machine

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* Address feedback

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* Use new_native_or_wasm_executor

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chain_head: Remove 'static on Backend

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chain_head: Add documentation

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chain_head: Lock blocks before async blocks

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chain_head_follower: Remove static on backend

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* Update client/service/src/builder.rs

Co-authored-by: Davide Galassi <davxy@datawok.net>

* Update client/service/src/builder.rs

Co-authored-by: Davide Galassi <davxy@datawok.net>

* chain_head: Add BlockHeaderAbsent to the PartialEq impl

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* client: Add better documentation around pinning constants

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* chain_head: Move subscription to dedicated module

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

* subscription: Rename global pin / unpin functions

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>

---------

Signed-off-by: Alexandru Vasile <alexandru.vasile@parity.io>
Co-authored-by: Bastian Köcher <git@kchr.de>
Co-authored-by: parity-processbot <>
Co-authored-by: Sebastian Kunert <skunert49@gmail.com>
Co-authored-by: Davide Galassi <davxy@datawok.net>

juangirini committed:
contracts: Make Origin information available (#13708)

* contracts: allow root calls

* contracts: update ContractOrigin

* contracts: test allow root calls

* contracts: rustfmt

* contracts: test root caller traps

* contracts: add Caller enum

* contracts: improve some comments

* contracts: improve some comments

* contracts: format code with +nightly

* contracts: fix failing tests

* contracts: improve charte instantiate call when root origin

* contract: refactor common, call and instantiate inputs

* contracts: fix some failing test

* contracts: trap caller when there is no account id

* Update frame/contracts/src/lib.rs

Co-authored-by: PG Herveou <pgherveou@gmail.com>

* Update frame/contracts/src/exec.rs

Co-authored-by: PG Herveou <pgherveou@gmail.com>

* contracts: make `into_deposit` return zero when the origin is root

* contracts: cargo fmt

* contracts: charging and terminating tests refactored

* contracts: improved errors

* contracts: refactor & merge ContractOrigin cand Caller enums

* Update frame/contracts/src/lib.rs

Co-authored-by: Sasha Gryaznov <hi@agryaznov.com>

* contracts: apply suggested pr changes

* contracts: rename Caller to Origin

* contracts: refactor run fn

* contracts: cr improvements

* contracts: allow root to use delegate call

* contracts: add caller_is_root weight

* ".git/.scripts/commands/bench/bench.sh" pallet dev pallet_contracts

* contracts: add caller_is_root benchmarking

* ".git/.scripts/commands/bench/bench.sh" pallet dev pallet_contracts

* contracts: add caller_is_root benchmarking

* ".git/.scripts/commands/bench/bench.sh" pallet dev pallet_contracts

* contracts: add some minor improvements

* contracts: make caller_is_root runtime fn unstable

* contracts: fix failing wasm test

* contracts: update benchmarking for origin_is_root

* contracts: improve seal_caller_is_root benchmarking

* ".git/.scripts/commands/bench/bench.sh" pallet dev pallet_contracts

* contracts: add small pr improvements

* contracts: fix broken tests after master merge

* contracts: acknowledge the default storage deposit limit

* ".git/.scripts/commands/bench/bench.sh" pallet dev pallet_contracts

* contracts: move origin to CommonInput

* contracts: add some extra tests

* contracts: move ensure origin logic inside invokable::run_guarded

* contracts: add minor improvements

* contracts: fix caller_is_root benchmarking

* contracts: improve function description

* Update frame/contracts/src/lib.rs

Co-authored-by: Sasha Gryaznov <hi@agryaznov.com>

* Update frame/contracts/src/lib.rs

Co-authored-by: Sasha Gryaznov <hi@agryaznov.com>

* Update frame/contracts/src/wasm/runtime.rs

Co-authored-by: Sasha Gryaznov <hi@agryaznov.com>

* contracts: improve function description

---------

Co-authored-by: parity-processbot <>
Co-authored-by: PG Herveou <pgherveou@gmail.com>
Co-authored-by: Sasha Gryaznov <hi@agryaznov.com>

Commit:
Bump clap to 4.2.5 (#14061)

Co-authored-by: parity-processbot <>

shunsukew committed:
Manual seal delayed finalize (#13999)

* up

* up

* added test

* remove unncessary dep

* cargo fmt

* cargo fmt

* up

* Update client/consensus/manual-seal/src/lib.rs

Co-authored-by: Bastian Köcher <git@kchr.de>

* fix test

* cargo fmt

* added docs

* updated doc

---------

Co-authored-by: Bastian Köcher <git@kchr.de>
Co-authored-by: parity-processbot <>

Commit:
Don't run `check-crates-publishing` job on prs (#14064)

* Restore `check-crates-publishing-pipeline` job

This job was errorneously deleted in lieu of the
`check-crates-publishing` job.

This reverts commit de80d0107336a9c7a2efdc0199015e4d67fcbdb5.

* Don't run `check-crate-publishing` job on prs.

Originally attempted in #14044. It's been broken for a while and we're
not sure yet how to move forward in the longer term.

skunert committed:
Only calculate tree route during finalization when there are multiple leaves (#14067)

* Only calculate tree route when there are multiple leaves

* Update client/service/src/client/client.rs

Co-authored-by: Bastian Köcher <git@kchr.de>

---------

Co-authored-by: Bastian Köcher <git@kchr.de>

bkchr committed:
test-staking-e2e: Add to main `Cargo.toml`. (#14062)

This ensures that it is actually tested as part of the CI. There is also no need to have it separate.

juangirini committed:
contracts: add events to ContractResult (#13807)

* contracts: add events to ContractResult

* contracts: add encoded events to ContractResult

* contracts: add generic Event to ContractResult

* contracts: test bare_call events

* contracts: update bare_call test name

* contracts: add better comments to dry run events

* contracts: fix pallet contracts primitives implementation

* contracts: add EventRecord generic to ContractInstantiateResult

* contracts: make event collection optional

* contracts: impreved notes on `collect_events`

* contracts: update benchmarking calls

* contracts: change bare_call and bare_instantiate to accept enums instead of bools

* contracts: add partial eq to new enums

* contracts: improve comments

* contracts: improve comments

* contracts: fix bare_call benchmarking

* contracts: fix bare call and instantiate in impl_runtime_apis

* contracts: add api versioning to new ContractsApi functions

* contracts: modify api versioning to new ContractsApi functions

* contracts: create new contracts api trait

* contracts: clean up code

* contracts: remove the contract results with events

* contracts: undo contracts api v3

* contracts: remove commented out code

* contracts: add new test bare call result does not return events

* contracts: add code review improvements

* contracts: remove type imports

* contracts: minor code review improvements

niklasad1 committed:
rpc server: break legacy CLI options and remove "backward compatible HTTP server" (#13384)

* jsonrpsee v0.16

* breaking: remove old CLI configs

* remove patch.crates-io

* fix bad merge

* fix clippy

* fix bad merge

* fix grumbles

* Update client/service/src/lib.rs

Co-authored-by: Bastian Köcher <git@kchr.de>

* revert block_in_place

* add issue link in todo

* Update client/cli/src/config.rs

Co-authored-by: Dmitry Markin <dmitry@markin.tech>

* grumbles: add ipv6 loopback address

* Revert "grumbles: add ipv6 loopback address"

This reverts commit 3a0b1ece6c4e36055d666896c29d1da55ffa1c4f.

* remove nits

* bump zombienet version

* adress grumbles: provide structopt default_val_t

* remove duplicate from structopt

* bump zombienet v1.3.47

* bump zombienet version

---------

Co-authored-by: Bastian Köcher <git@kchr.de>
Co-authored-by: Dmitry Markin <dmitry@markin.tech>
Co-authored-by: Javier Viola <javier@parity.io>

gpestana committed:
Update frame/election-provider-multi-phase/src/signed.rs

Co-authored-by: Ankan <10196091+Ank4n@users.noreply.github.com>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2777735

gpestana committed:
Update bin/node/runtime/src/lib.rs

Co-authored-by: Ankan <10196091+Ank4n@users.noreply.github.com>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2777892

gpestana committed:
Makes DepositCalculator trait more generic

gpestana committed:
Merge branch 'master' into gpestana/epm_variable_deposit_base

gpestana committed:
Fixed runtime configs

gpestana commented:
bot rebase

Commit:
Merge remote-tracking branch 'origin/master' into gpestana/epm_variable_deposit_base

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2778941

gpestana committed:
Updates e2e integration tests mock

kianenigma commented on frame/election-provider-multi-phase/src/signed.rs:
```diff
@@ -349,6 +351,39 @@ impl<T: Config> SignedSubmissions<T> {
+}
+
+impl<T: Config> DepositCalculator<BalanceOf<T>> for GeometricDepositBase<T> {
+	/// The fixed deposit base upon which the geometric series is calculated.
```
I don't see a need for these to be associated types; can you expand why? you can use `T::SignedDepositBaseIncreaseFactor` and so on directly.

These had to be associated types iff the input to these pallet was only a singular `type Deposit: DepositCalculator<Balance>`. Then all aspects of the deposit, including weight, len, and queue length need to be abstracted away and compacted into a single `T::Deposit::calculate(queue_len, solution_weight, solution_len)`. 

In either case, the existing `type BaseDeposit` and `type IncreaseFactor` don't serve any purpose.

kianenigma replied:
An alternative design that seems clean to me is to have: 

```
trait Config {
  type DepositLen: Convert<usize, Balance>;
  type DepositWeight: Convert<weight, Balance>;
  type DepositByte: Convert<u32, Balance>;
}
```

each in charge of calculating one dimension of the deposit. The final deposit would be the sum of two. The previous suggestion was merely hiding all 3 behind a single function and a custom trait instead of `Convert`. Albeit, even for the previous one, you could do `Convert<(usize, u32, Weight), Balance>`, but that's a bit too ugly.

gpestana replied:
> I don't see a need for these to be associated types; can you expand why? you can use T::SignedDepositBaseIncreaseFactor and so on directly.

The goal was to keep the `trait DepositCalculator` as general as possible and to provide the implementors a way to fetch the `IncreaseFactor` and the `BaseDeposit` externally. I agree that these types are redundant when the implementor has access to `T` though, as it is in this implementation.

I will rethink this trait.

gpestana replied:
So I think a good compromise which is clean IMO is to keep the `SIgnedDepositByte` and `SignedDepositWeight` as is (i.e. no refactor to use `Convert<A, Balance>`) and keep the current implementation of `fn deposit_for` returning the final deposit balance.

However, the I have removed the previous `trait DepositCalculator ` (which I agree was a bit convoluted) and added a new associated type to EPM's config that converts a usize to Balance. And implement a geometric progression on a type that implements convert.

kianenigma commented on frame/election-provider-multi-phase/src/signed.rs:
```diff
@@ -349,6 +351,39 @@ impl<T: Config> SignedSubmissions<T> {
+	/// The nth progression term will be `IncreaseFactor`% larger than the nth-1 term.
+	type IncreaseFactor = T::SignedDepositBaseIncreaseFactor;
+
+	/// Increase the base deposit as a geometric progression based on the number of signed
```
nit: implementations can be commented with just `//`, not `///`. See https://hackmd.io/mRrrewoxTOyfKJ2H16R2xw?both, to be made public soon.

kianenigma commented on frame/election-provider-multi-phase/src/lib.rs:
```diff
@@ -673,6 +675,16 @@ pub mod pallet {
 
+		/// Something that calculates the signed deposit base based on the signed submissions queue
+		/// size.
+		type SignedDepositBase: DepositCalculator<BalanceOf<Self>>;
```
What is `SignedFixedDepositBase` and `SignedDepositBase`?

gpestana replied:
`SignedDepositBase` is the type that implements the final signed deposit base (i.e the `DepositCalculator` implementor). `SignedFixedDepositBase`, on the other hand, is the initial term of the progression (thus "Fixed").

Simplifying, we have:

```
fn signed_deposit_base(n_term: usize) -> Balance {
  signed_fixed_deposit_base * (1 + increase_factor)^n_term
} 
```

I agree that the naming is not ideal though. I think that with the simplifications that you suggested above, we can improve on the naming as well.

gpestana replied:
I improved the naming, I think it's looking better and more clear now.

kianenigma reviewed (commented):
I think there's some room for improvement with the traits here.

Commit:
Merge remote-tracking branch 'origin/master' into gpestana/epm_variable_deposit_base

gpestana committed:
Refactors the traits

Ank4n commented on frame/election-provider-multi-phase/src/signed.rs:
```diff
@@ -349,6 +351,31 @@ impl<T: Config> SignedSubmissions<T> {
+
+impl<T: Config> Convert<usize, BalanceOf<T>> for GeometricDepositBase<T> {
+	// Calculates the base deposit as a geometric progression based on the number of signed
+	// su Calculates.
```
```suggestion
	// submissions.
```

Ank4n commented on frame/election-provider-multi-phase/src/signed.rs:
```diff
@@ -349,6 +351,31 @@ impl<T: Config> SignedSubmissions<T> {
+	// 5, with initial deposit of 1000 and 10% of increase factor is 1000 * (1 + 0.1)^5.
+	fn convert(queue_len: usize) -> BalanceOf<T> {
+		let increase_factor: FixedU128 =
+			<Percent as Into<FixedU128>>::into(T::SignedDepositBaseIncreaseFactor::get()) +
+				sp_runtime::traits::One::one();
```
This is a little more concise way to do it.
```suggestion
			FixedU128::from_u32(1) + T::SignedDepositBaseIncreaseFactor::get().into();
```

Ank4n reviewed (approved):
Couple of nits, otherwise LGTM!

gpestana committed:
Update frame/election-provider-multi-phase/src/signed.rs

Co-authored-by: Ankan <10196091+Ank4n@users.noreply.github.com>

kianenigma reviewed (commented):
I think this is mainly done, but needs to be brought up to date before a final review.

gpestana committed:
Nits and addresses review comments

gpestana committed:
Merge branch 'master' into gpestana/epm_variable_deposit_base

gpestana commented on bin/node/runtime/src/lib.rs:
```diff
@@ -747,7 +748,9 @@ impl pallet_election_provider_multi_phase::Config for Runtime {
 	type SignedMaxSubmissions = ConstU32<10>;
 	type SignedRewardBase = SignedRewardBase;
-	type SignedDepositBase = SignedDepositBase;
+	type SignedDepositBase = GeometricDepositBase<Runtime>;
```
It is, yes. `SignedDepositBase` must implement the `Convert<usize, BalanceOf<T>>` trait. `GeometricDepositBase<T>` implements the convert trait with the geometric progression of the deposit based on the size of the submission queue.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2962816

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2962817

gpestana committed:
Merge branch 'master' into gpestana/epm_variable_deposit_base

gpestana commented:
bot fmt

Commit:
".git/.scripts/commands/fmt/fmt.sh"

kianenigma commented on frame/election-provider-multi-phase/src/lib.rs:
```diff
@@ -647,9 +648,16 @@ pub mod pallet {
+		/// If 0, then signed base deposit remains constant regardless of the submissions queue size
+		/// (and equal to `SignedFixedDeposit`).
+		#[pallet::constant]
+		type SignedDepositIncreaseFactor: Get<Percent>;
```
This associated type is not useful if `SignedDepositBase` is not implementing a geometric distribution. Same can be said for the type above. These should be moved to `GeometricDepositBase`: 

```rust
struct GeometricDepositBase<Balance, Base: Get<Balance>, Inc: Get<Percent>>(...)
```

gpestana replied:
Great suggestion, updated the PR.

gpestana committed:
Parameterize GeometricProgression instead of using dedicated associated types for progression configs

kianenigma commented:
can be merged once conflict is resolved.

gpestana committed:
Merge branch 'master' into gpestana/epm_variable_deposit_base
